index: Deduplicate HashKey / HeightKey handling

mzumsande opened the pull request:
The logic for `DBHashKey` / `DBHeightKey` handling and lookup of entries is shared by `coinstatsindex` and `blockfilterindex`, leading to many lines of duplicated code. De-duplicate this by moving the logic to `index/db_key.h` (using templates for the index-specific `DBVal`).

mzumsande committed:
index: deduplicate Hash / Height handling

The code was largely duplicated between coinstatsindex
and blockfilterindex.
Deduplicate it by moving it to a shared file.

slight change in behavior: the index name is no longer
part of the error msg in case of (un)serialization errors.

mzumsande committed:
index, refactor: deduplicate LookUpOne

LookUpOne is used by both coinstatsindex and blockfilterindex,
the two implementations had already started to deviate slightly
for no apparent reason.

fjahr commented on src/index/coinstatsindex.cpp:
```diff
@@ -119,23 +119,6 @@ bool CoinStatsIndex::CustomAppend(const interfaces::BlockInfo& block)
     // Ignore genesis block
     if (block.height > 0) {
-        std::pair<uint256, DBVal> read_out;
-        if (!m_db->Read(DBHeightKey(block.height - 1), read_out)) {
```
It's not true that they aren't used for anything. The point of reading the block from disk is to check if it's the prev block that is expected because the values that are written for the current block depend on the prev block values. They are used not from the prev block but from the members instead though. I am not sure right now if this check can just be removed, but if they are still needed the expected prev block hash could probably be written into a member too and the DB read could be avoided this way.

fjahr replied:
Yes, the return false would prevent potentially writing false stats because the member variables are not up-to-date with what the prev block should be. I don't remember what the scenarios were but I think there was also fear of some edge cases with reorgs but from today's perspective I am not sure anymore how that would happen. Maybe the refactorings in the index have made this scenario impossible but I'll have to think about it some more.

mzumsande replied:
I'm skeptical if this is a possible scenario with today's index base code (which is much better with reorgs / unexpected restart scenarios than it was back then), but it's definitely more complicated than I thought, so this PR is not appropriate for this - I dropped the commit and will resolve this discussion.

furszy replied:
The base index class is responsible for detecting reorgs and calling the appropriate methods from the child class: `CustomRemove()` up to the forking point, then `CustomAppend()` to connect the new chain blocks. This happens during both initial sync and the validation event. I'm pretty confident we can safely remove the extra check.

Furthermore, since the index child class receives block connection and disconnection events in order, I don't see why it should care about reorgs at all (speaking generally here, for all index child classes). That's one of the base class responsibilities.

Also, if we do want to keep the previous block check for some reason, I think we could cache it during `CustomInit()`, then update it in both `CustomAppend()` and `CustomRemove()`. Or.. we could keep the base class’s `m_best_block_index` updated with every processed block too. Either way, I think we can avoid accessing disk for this.

fjahr commented:
Concept ACK on the deduplication parts

fjahr reviewed (commented):
Looks good, aside from the documentation cleanup in `blockfilterindex.cpp` you can consider my comments optional.

mzumsande commented on src/index/key.h:
```diff
@@ -76,4 +76,24 @@ template <typename DBVal>
 }
 
+template <typename DBVal>
+static bool LookUpOne(const CDBWrapper& db, const interfaces::BlockRef& block, DBVal& result)
```
added these and `serialize.h`, hope that'll do. (never got iwyu running locally and the tidy job didn't cover the new file for some reason).

mzumsande commented:
[03b67a2](https://github.com/bitcoin/bitcoin/commit/03b67a27d013e55891df69b1aef9425574b88f5e) to [936bd87](https://github.com/bitcoin/bitcoin/commit/936bd874ac1581de532317741f5ab26e875935b3):
Addressed feedback by @fjahr.

fjahr commented:
utACK 936bd874ac1581de532317741f5ab26e875935b3

furszy reviewed (commented):
Since `DBHeightKey` and `DBHashKey` are pretty similar. What do you think about a generic template for both? https://github.com/furszy/bitcoin-core/commit/e5dd2b02f15f9f68fbe84cc30255573ad23dbf2c.
(Only needed to specialize the serialization for the value)

furszy reviewed (commented):
what do you think about wrapping all the code in `db_key.h` under a namespace like "util", "common", "index::util", or "util::index" (just threw some names), so it's clear when reading the code that these functions are defined elsewhere and shared, without needing to search around for the definition.

mzumsande committed:
index: restrict index helper function to namespace

mzumsande commented:
> what do you think about wrapping all the code in `db_key.h` under a namespace like "util", "common", "index::util", or "util::index" (just threw some names), so it's clear when reading the code that these functions are defined elsewhere and shared, without needing to search around for the definition.

I did that in 5646e6c0d3581f12419913b88745f51c7a3161b9 - let me know what you think. I used `index_util`, turned out it was required a bit more often than I would have expected.

furszy reviewed (commented):
utACK 5646e6c0d3581f12419913b88745f51c7a3161b9

sedited reviewed (approved):
ACK 5646e6c0d3581f12419913b88745f51c7a3161b9

Could run `clang-format-diff` on the commits.
